monash-human-power/MHP_DAS
# Implement crash alert

## Description

Implements the crash alert system that allows anyone subscribed to receive an alert via their device when the bike crashes.

Adds code that listens for messages from the crash detection and upon receiving a `true`, sends an alert via [this Slack channel](https://monashhumanpower.slack.com/archives/C041XHE6MEJ) in the MHP Slack workspace. This can be extended for more APIs by adding children of the `MessageSender` class and implementing the `send_message()` method with the APIs respective calls.

## Screenshots

Message in the Slack channel when a crash is detected:
<img width="390" alt="image" src="https://user-images.githubusercontent.com/90585684/206054591-0d125b64-fd30-45f8-87f1-ceefce578ccb.png">


## Steps to Test

1. Create/activate your poetry environment using python 3.7. (i.e. `poetry env use 3.7`)
2. Spawn into the environment using `poetry shell`
3. Install the dependencies with `poetry install`
4. Run the main program using `python main_crash_alert.py`

The program should start listening to the MQTT channel:
<img width="765" alt="image" src="https://user-images.githubusercontent.com/90585684/206058743-c06af2fb-fd12-43e1-bea9-fced03ff5e21.png">

5. In another window, publish to the MQTT channel:
```
# to send a false
mosquitto_pub -t "/v3/wireless_module/3/crash_detection" -m '{"value": false}'

# to send a true
mosquitto_pub -t "/v3/wireless_module/3/crash_detection" -m '{"value": true}'
```
When a `true` is sent, a message should be sent in the Slack channel if it has been 10 seconds since the last `true` was received. When a `false` is received, no message should be sent.
